nemo/core/callbacks.py

```diff
@@ -35,6 +54,463 @@
+                    raise NotImplementedError(
+                        "There were two instances of the same module. Please overwrite __str__() of one of the "
+                        "modules."
+                    )
+                unique_mod_names.add(str(module))
+                if self._step_freq > -1:
+                    filename = f"{module}-STEP-{state['step']}.pt"
+                else:
+                    filename = f"{module}-EPOCH-{state['epoch']}.pt"
+                module.save_to(os.path.join(path, filename))
+
+        if self._step_freq > -1:
+            filename = f"trainer-STEP-{state['step']}.pt"
+            state.save_state_to(f"{path}/{filename}")
+            self._saved_ckpts.append(f"-{state['step']}.pt")
+        else:
+            filename = f"trainer-EPOCH-{state['epoch']}.pt"
+            state.save_state_to(f"{path}/{filename}")
+            self._saved_ckpts.append(f"-{state['epoch']}.pt")
+
+        if len(self._saved_ckpts) > self._ckpt2keep:
+            for end in self._saved_ckpts[: -self._ckpt2keep]:
+                for file in glob.glob(f'{path}/*{end}'):
+                    os.remove(file)
+            self._saved_ckpts = self._saved_ckpts[-self._ckpt2keep :]
+        logging.info(f'Saved checkpoint: {path}/{filename}')
+
+    def __restore_from(self, path, state):
+        if not os.path.isdir(path):
+            if self._force_load:
+                raise ValueError("force_load was set to True for checkpoint callback but a checkpoint was not found.")
+            logging.warning(f"Checkpoint folder {path} not found!")
+        else:
+            logging.info(f"Found checkpoint folder {path}. Will attempt to restore checkpoints from it.")
+            modules_to_restore = []
+            modules_to_restore_name = []
+            for module in AppState().modules:
+                if module.num_weights > 0:
+                    modules_to_restore.append(module)
+                    modules_to_restore_name.append(str(module))
+            try:
+                module_checkpoints = get_checkpoint_from_dir(modules_to_restore_name, path)
+
+                for mod, checkpoint in zip(modules_to_restore, module_checkpoints):
+                    mod.restore_from(checkpoint, state["local_rank"])
+            except (ValueError) as e:
+                if self._force_load:
+                    raise ValueError(
+                        "force_load was set to True for checkpoint callback but a checkpoint was not found."
+                    )
+                logging.warning(e)
+                logging.warning(
+                    f"Checkpoint folder {path} was present but nothing was restored. Continuing training from random "
+                    "initialization."
+                )
+                return
+
+            try:
+                trainer_checkpoints = get_checkpoint_from_dir(["trainer"], path)
+                state.restore_state_from(trainer_checkpoints[0])
```

Comment:
ok, I thin we should explain the format of this checkpoint

It seems that state (from StateWrapper?) is in [0], what else is down there...?